# Programa que calcula la lista de numeros b tales que b es un multiplicador de periodo maximo para un MCG de modulo 541

In [20]:
2**8+2**3+2**2+2**1

270

## Esta primera función toma un entero b entre 2 y 540 y determina si alguno de los residuos de b^108, b^180 o b^270 modulo 541 son iguales a 1, en caso afirmativo devuelve False y en caso contrario devulve True

In [21]:
def residuo(b):
    rb2=(b*b)%541#residuo de b^2 modulo 2023
    rb4=(rb2*rb2)%541#residuo de b^(2^2) modulo 2023
    rb8=(rb4*rb4)%541#residuo de b^(2^3) modulo 2023
    rb16=(rb8*rb8)%541#residuo de b^(2^4) modulo 2023
    rb32=(rb16*rb16)%541#residuo de b^(2^5) modulo 2023
    rb64=(rb32*rb32)%541#residuo de b^(2^6) modulo 2023
    rb128=(rb64*rb64)%541#residuo de b^(2^7) modulo 2023
    rb256=(rb128*rb128)%541#residuo de b^(2^8) modulo 2023
    rb108=(rb64*rb32*rb8*rb4)%541#residuo de b^(2^6+2^5+2^3+2^2) modulo 2023
    if rb108==1:
        return False
    rb180=(rb128*rb32*rb16*rb4)%541#residuo de b^(2^7+2^5+2^4+2^2) modulo 2023
    if rb180==1:
        return False
    rb270=(rb256*rb8*rb4*rb2)%541#residuo de b^(2^1+2^2+2^3+2^8) modulo 2023
    if rb270==1:
        return False
    else:
        return True    

## Ahora buscamos y almacenamos todos los enteros mayores a 1 y menores a 541 que son primos relativos a 2023 y que NO devuelven 1 en la funcion residuo

In [24]:
from math import gcd

In [22]:
def encontrar():
    b=[];
    k=2
    while k<541:
        if gcd(k,541)==1:
            if residuo(k):
                b.append(k)
        k=k+1
    return b

In [25]:
multipliers=encontrar()

In [26]:
multipliers

[2,
 10,
 13,
 14,
 18,
 24,
 30,
 37,
 40,
 51,
 54,
 55,
 59,
 62,
 65,
 67,
 68,
 72,
 73,
 77,
 83,
 86,
 87,
 91,
 94,
 96,
 98,
 99,
 107,
 113,
 114,
 116,
 117,
 126,
 127,
 128,
 131,
 132,
 138,
 150,
 152,
 153,
 156,
 158,
 163,
 176,
 181,
 183,
 184,
 195,
 197,
 199,
 206,
 208,
 210,
 213,
 218,
 220,
 223,
 224,
 244,
 248,
 250,
 257,
 258,
 259,
 260,
 261,
 263,
 267,
 269,
 270,
 271,
 272,
 274,
 278,
 280,
 281,
 282,
 283,
 284,
 291,
 293,
 297,
 317,
 318,
 321,
 323,
 328,
 331,
 333,
 335,
 342,
 344,
 346,
 357,
 358,
 360,
 365,
 378,
 383,
 385,
 388,
 389,
 391,
 403,
 409,
 410,
 413,
 414,
 415,
 424,
 425,
 427,
 428,
 434,
 442,
 443,
 445,
 447,
 450,
 454,
 455,
 458,
 464,
 468,
 469,
 473,
 474,
 476,
 479,
 482,
 486,
 487,
 490,
 501,
 504,
 511,
 517,
 523,
 527,
 528,
 531,
 539]

In [27]:
len(multipliers)

144